### ISSUE TITLE
VoxelTool::do_box for smooth terrain

### ISSUE THREAD

**Issue by blockspacer** [Zylann/godot_voxel#109]
I tried to implement `VoxelTool::do_box` and noticed that produced shape has rounded corners

```cpp
void VoxelTool::do_box(Vector3i begin, Vector3i end) {
	Rect3i box(begin, end - begin);

	ERR_FAIL_COND(end.distance_sq(begin) <= 0);

	if (!is_area_editable(box)) {
		print_line("Area not editable");
		return;
	}
	box.for_each_cell([this](Vector3i pos) {
		_set_voxel_f(pos, _mode == MODE_REMOVE ? 100.0 : -100.0);
	});

	_post_edit(box);
}
```

I suspect that it is due `Sharp Features` not supported by transvoxel https://github.com/Zylann/godot_voxel/issues/24

Is it possible to enable `Sharp Features` without rewriting a lot of `godot_voxel` logic, for example, extend transvoxel?

Can we introduce a new channel in the voxel data that stores an 8-bit sharpness parameter similar to approach from https://transvoxel.org/Lengyel-VoxelTerrain.pdf ?

> Several methods have been developed for capturing sharp features during polygonalization,
> and they include Extended Marching Cubes [Kobbelt et al. 2001], Dual Marching Cubes algorithm [Nielson 2004], and Dual Contouring [Ju et al. 2002]. Each of these techniques involves
> placing extra vertices in the interior of a cell in order to generate edges having a smaller dihedral
> angle. Using similar methods, the per-face edge configurations shown in Figure 3.4 are transformed into those shown in Figure 6.1, and these would be applied to triangulations for both regular cells and transition cells. The positions of the interior vertices are determined by the normal
> vectors at the vertices lying on the cell edges, as shown in Figure 6.2. In order to create a sharp
> edge while shading, interior vertices would ordinarily be duplicated and assigned different normal
> vectors corresponding to their adjacent faces. Whether this duplication occurs can be controlled
> by setting a dihedral angle threshold above which a smoothed normal vector is used instead. This
> prevents faceting on smoothly curving surfaces while also allowing sharp edges to be produced. It
> is also possible to eliminate the interior vertex altogether if the dihedral angle is close to 180 degrees, as in the leftmost example in Figure 6.2.
>  So that an artist could choose whether smooth or sharp triangulations are produced on a percell basis, we could introduce a new channel in the voxel data that stores an 8-bit sharpness parameter. A value of 0 would indicate a smooth surface generated without any interior vertices,
> and a value of 255 would indicate the maximum sharpness. For intermediate values, the positions
> of interior vertices would be interpolated between the position for maximum sharpness and its
> projection onto the edge that would be created without an interior vertex. Sharpness would need
> to be determined independently for each edge based on interpolated sharpness values stored for
> each corner of a cell, and this would greatly increase the number of triangulations that could be
> generated for each equivalence class.

**Comment by tinmanjuggernaut** [comment]
You're running into a limitation of transvoxel.

Rather than extending Transvoxel for a likely slower, poor result, it would be better to implement a more accurate, faster mesher such as MDC, DC, surface nets. See #24 

UE4 plugin went with surface nets.

**Comment by blockspacer** [comment]
MDC looks good
What do you think about **manifold dual contouring**?
(sources)
https://github.com/GarryGuanDoubleG/Unity-Voxel-Plugin/blob/d0ba469033136899e6513469d95027fd5c58ba95/README.md#L3
https://github.com/Russoul/Voxelized3D-Rust/blob/master/src/extraction/uniform_manifold_dc.rs#L15
(Performance Comparison)
https://github.com/Visse/portfolio/blob/38ce87ed8b80acb61e204d692c43ad81657bdb61/volume/index.md#performance-comparison
https://github.com/Lin20/isosurface
https://github.com/Svenny/isomesh-viewer/blob/22e9cc6e790d412b67a1299d781be12e069505a0/src/mainwindow.cpp#L334
(papers)
https://github.com/Visse/portfolio/tree/38ce87ed8b80acb61e204d692c43ad81657bdb61/volume/res

**Comment by tinmanjuggernaut** [comment]
Unique references should be added to #24. I've already added Lin20's implementation and shared my thoughts on that ticket. DC variations have less vertices == faster, and represent corners better.

Nevertheless, it's good to implement do_box as that is independent of the mesher and applicable to blocky as well. You should create a PR for it.

**Comment by blockspacer** [comment]
@tinmanjuggernaut 

I tried to use `VoxelMesherDMC` and `VoxelTool::do_box` (code above) produces sliced off corners

![image](https://user-images.githubusercontent.com/35254900/73139579-6ef89d80-4080-11ea-89c1-04143bbf170a.png)
Image from Dual Contouring tutorial https://www.boristhebrave.com/2018/04/15/dual-contouring-tutorial/

Is it possible to turn on `Dual Contouring` in `VoxelMesherDMC`?

**Comment by tinmanjuggernaut** [comment]
> Is it possible to turn on sharp edges in `VoxelMesherDMC`?

With the algorithm as it is, no. MC, DMC, Transvoxel do not support sharp edges. You need an algorithm that does, such as dual contouring and variations. The current algorithm could be modified, but that effort would be better spent creating a new mesher with a better algorithm, IMO. I considered implementing MDC, but decided to spend my time on other projects like implementing FastNoise in Godot. You can take a shot at it.

**Comment by blockspacer** [comment]
> > Is it possible to turn on sharp edges in `VoxelMesherDMC`?
> 
> With the algorithm as it is, no. MC, DMC, Transvoxel do not support sharp edges. You need an algorithm that does, such as dual contouring and variations. The current algorithm could be modified, but that effort would be better spent creating a new mesher with a better algorithm, IMO. I considered implementing MDC, but decided to spend my time on other projects like implementing FastNoise in Godot. You can take a shot at it.

I'll be happy to participate in DMC implementation for godot_voxel, but now i don't have enough knowledge about DMC to implement it as sole contributor, without asking a LOT of questions. 

One solution to that problem - i can try to port existing C++ code from https://github.com/Svenny/isomesh-viewer/blob/22e9cc6e790d412b67a1299d781be12e069505a0/src/mainwindow.cpp#L334 to godot_voxel. I can't guarantee that attempt will be successful. Also ported code may require code review cause godot_voxel uses custom threading model

What do you think about `AlgoDualContouring` and `AlgoManifoldDualContouring` from https://github.com/Svenny/isomesh-viewer ? Is it good idea to port that https://github.com/Svenny/isomesh/blob/master/src/algo/uniform_dual_contouring.cpp into godot_voxel?

**Comment by tinmanjuggernaut** [comment]
> i don't have enough knowledge about DMC to implement it as sole contributor

Welcome to the club. 

> without asking a LOT of questions

Zylann can answer questions about Voxel Tools and what he did with DMC and Transvoxel. I probably can't answer much in this regard.

> One solution to that problem - i can try to port existing C++ code from...

If you're going to do it, you'll have to decide which implementation. It looks like there are two implementations of manifold dual contouring. This one and Lin20's. Or you can read the research paper itself. 

I personally wouldn't bother implementing an algorithm with known issues like being non-manifold or has rough edges, or creates more triangles than DMC.

**Comment by Zylann** [comment]
This was implemented by #172
